File: codes/Tree/110.balanced-binary-tree/balanced-binary-tree.go
package main


type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isBalanced(root *TreeNode) bool {
	if root==nil {
		return true
	}
	if countHeight(root)==-1{
		return false
	}
	return true
}

func countHeight(node *TreeNode) int {
	if node == nil {
		return 0
	}
	leftHeight := countHeight(node.Left)
	if leftHeight == -1 {
		return -1
	}
	rightHeight := countHeight(node.Right)
	if rightHeight == -1 {
		return -1
	}

	if abs(leftHeight-rightHeight) > 1 {
		return -1
	}
	return max(leftHeight, rightHeight) + 1
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a

}

func max(a,b int)int{
	if a<b{
		return b
	}
	return a
}

File: codes/Tree/110.balanced-binary-tree/balanced-binary-tree.cpp
#include <stack>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {
    }
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {
    }
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {
    }
};

class Solution {
public:
    bool isBalanced(TreeNode *root) {
        if (!root) {
            return true;
        }
        return getHeight(root) == -1 ? false : true;
    }
    int getHeight(TreeNode *node) {
        if (!node) {
            return 0;
        }
        auto leftHeight = getHeight(node->left);
        if(leftHeight==-1){
            return -1;
        }
        auto rightHeight = getHeight(node->right);
        if(rightHeight==-1){
            return -1;
        }
        return abs(leftHeight - rightHeight) > 1 ? -1 : max(leftHeight, rightHeight) + 1;
    }
};